### PR TITLE
Add ability to set a custom fallback translator

### PR DESCRIPTION
**Pull request by andrej-griniuk**
Refs #9257

### CODE REVIEW

File: src/I18n/I18n.php [hunk 1]
```diff
@@ -201,6 +201,18 @@ public static function config($name, callable $loader)
     }
 
     /**
+     * Registers a custom fallback translator loader
+     *
+     * @param callable $loader A callable object that should return a Package
+     * instance to be used for assembling a new translator.
+     * @return void
+     */
+    public static function fallbackConfig(callable $loader)
```
Comment (markstory) [inline src/I18n/I18n.php]: Any reason to not use `setFallbackLoader()`?

File: src/I18n/I18n.php [hunk 2]
```diff
@@ -201,6 +201,18 @@ public static function config($name, callable $loader)
...
+    {
+        static::translators()->setCustomFallbackLoader($loader);
```
Comment (markstory) [inline src/I18n/I18n.php]: Same question here `setFallbackLoader()`?

File: src/I18n/TranslatorRegistry.php [hunk 3]
```diff
@@ -36,6 +36,13 @@ class TranslatorRegistry extends TranslatorLocator
     protected $_loaders;
 
     /**
+     * Custom fallback loader
+     *
+     * @var callable
+     */
+    protected $_customFallbackLoader;
```
Comment (markstory) [inline src/I18n/TranslatorRegistry.php]: Just `_fallbackLoader` is ok.

File: src/I18n/TranslatorRegistry.php [hunk 4]
```diff
@@ -185,6 +203,10 @@ public function useFallback($enable = true)
      */
     protected function _fallbackLoader($name, $locale)
     {
+        if ($loader = $this->_customFallbackLoader) {
```
Comment (markstory) [inline src/I18n/TranslatorRegistry.php]: Please don't assign variables inside conditionals.

File: src/I18n/TranslatorRegistry.php [hunk 5]
```diff
@@ -185,6 +203,10 @@ public function useFallback($enable = true)
...
+            return $loader($name, $locale);
+        }
```
Comment (markstory) [inline src/I18n/TranslatorRegistry.php]: Couldn't the existing fallback loader be 'set' when the registry is created?

File: src/I18n/I18n.php (hunk 1 above)
Comment (andrej-griniuk) [inline src/I18n/I18n.php]: From user's perspective this method does pretty much same thing as `I18n::config()`, just without specifying domain name (e.g. for all domains), so I thought `I18n::config()` and `I18n::fallbackConfig()` would make sense.

Comment (markstory) [inline src/I18n/I18n.php]: Fair point. Would it be reasonable to have:

```php
I18n::config('_fallback', $customLoader)
```

Then we don't need any new methods, and fallback isn't set in a different way than other loaders, it just has a special name.

Comment (andrej-griniuk) [inline src/I18n/I18n.php]: Sounds good.

**Comment by markstory** [comment]
Looking good to me. Any feedbak @lorenzo?

File: src/I18n/TranslatorRegistry.php [hunk 6]
```diff
@@ -33,7 +36,14 @@ class TranslatorRegistry extends TranslatorLocator
      *
      * @var array
      */
-    protected $_loaders;
+    public $_loaders;
```
Comment (lorenzo) [inline src/I18n/TranslatorRegistry.php]: Why public?

Comment (andrej-griniuk) [inline src/I18n/TranslatorRegistry.php]: Ooops, left it there by accident.

File: src/I18n/TranslatorRegistry.php [hunk 7]
```diff
@@ -59,6 +69,48 @@ class TranslatorRegistry extends TranslatorLocator
     protected $_cacher;
 
     /**
+     *
+     * Constructor.
+     *
+     * @param PackageLocator $packages The package locator.
```
Comment (ADmad) [inline src/I18n/TranslatorRegistry.php]: FQCN required.

File: tests/TestCase/I18n/I18nTest.php [hunk 8]
```diff
@@ -511,6 +511,36 @@ public function testloaderFactory()
...
+     * Tests that it is possible to register a fallback translators factory
+     *
+     * @return void
+     */
+    public function testFallbackLoaderFactory()
+    {
+        I18n::config('_fallback', function ($name) {
+            $package = new Package('default');
+
+            if ($name == 'custom') {
```
Comment (dereuromark) [inline tests/TestCase/I18n/I18nTest.php]: Should be `===`

**Pull request merged**